docs/internals.rst

```diff
@@ -0,0 +1,314 @@
+
+* segment
+* offset
+
+The repository index files are random access but those files can be
+recreated if damaged or lost using ``check --repair``.
+
+Both indexes are stored as hash tables, directly mapped in memory from
+the file content, with only one slot per bucket, but that spreads the
+collisions to the following buckets. As a consequence the hash is just
+a start position for a linear search, and if the element is not in the
+table the index is linearly crossed until an empty bucket is
+found. When the table is full at 90% its size is doubled, when it's
+empty at 25% its size is halfed. So operations on it have a variable
+complexity between constant and linear with low factor, and memory
+overhead varies between 10% and 300%.
+
+The file chunk cache is stored in ``cache/files`` and is indexed on
+the ``file path hash`` and contains:
+
+* age
+* inode number
+* size
+* mtime_ns
+* chunks hashes
+
+The inode number is stored to make sure we distinguish between
+different files, as a single path may not be unique accross different
+archives in different setups.
+
+The file chunk cache is stored as a python associative array storing
+python objects, which generate a lot of overhead. This takes around
+240 bytes per file without the chunk list, to be compared to at most
+64 bytes of real data (depending on data alignment), and around 80
+bytes per chunk hash (vs 32), with a minimum of ~250 bytes even if
+only one chunck hash.
+
+Indexes memory usage
+--------------------
+
+Here is the estimated memory usage of |project_name| when using those
+indexes:
+
+Repository index
+  40 bytes x N ~ 200MB (If a remote repository is
+  used this will be allocated on the remote side)
+
+Chunk lookup index
+  44 bytes x N ~ 220MB
+
+File chunk cache
+  probably 80-100 bytes x N ~ 400MB
+
+Encryption
+----------
+
+AES_ is used with CTR mode of operation (so no need for padding). A 64
+bits initialization vector is used, a `HMAC-SHA256`_ is computed
+on the encrypted chunk with a random 64 bits nonce and both are stored
+in the chunk. The header of each chunk is : ``TYPE(1)`` +
+``HMAC(32)`` + ``NONCE(8)`` + ``CIPHERTEXT``. Encryption and HMAC use
+two different keys.
+
+In AES CTR mode you can think of the IV as the start value for the
+counter. The counter itself is incremented by one after each 16 byte
+block. The IV/counter is not required to be random but it must NEVER be
+reused. So to accomplish this Attic initializes the encryption counter
+to be higher than any previously used counter value before encrypting
+new data.
+
+To reduce payload size only 8 bytes of the 16 bytes nonce is saved in
+the payload, the first 8 bytes are always zeros. This does not affect
+security but limits the maximum repository capacity to only 295
+exabytes (2**64 * 16 bytes).
+
+Encryption keys are either a passphrase, passed through the
+``ATTIC_PASSPHRASE`` environment or prompted on the commandline, or
+stored in automatically generated key files.
+
```

Comment:
this might be a bit misleading, especially after the rather low-level description above it. it somehow sounds like the passphrase is used directly as a AES key. it is described better in the sections below, but still...